chainer/functions/array/where.py

```diff
@@ -16,30 +17,33 @@ def check_type_forward(self, in_types):
         type_check.expect(
             c_type.dtype == numpy.bool_,
             x_type.dtype == y_type.dtype,
-            x_type.shape == c_type.shape,
-            y_type.shape == c_type.shape,
         )
+        type_check.expect_broadcast_shapes(
+            c_type.shape, x_type.shape, y_type.shape)
 
     def forward(self, inputs):
+        # may broadcast
         self.retain_inputs((0,))
         xp = cuda.get_array_module(*inputs)
         condition, x, y = inputs
+        self.x_shape = x.shape
+        self.y_shape = y.shape
```

Comment:
You can get them in backward with `self.inputs[i].shape` (i=1,2), so no need to save them manually.